File: Tools/flashtool/WiFlyListener.py
# UDP broadcast receiver for Roving Networks/Microchip
# WiFly Wireless LAN Modules
#
import struct
import string
from datetime import datetime

from PyQt4.QtCore import *
from PyQt4.QtNetwork import *

BROADCAST_PORT = 55555


def zero_terminate(s):
    n = string.find(s, "\0")
    return s[0:n] if n > 0 else s


def unpack_info(data):
    info = {
        "ap_mac":       struct.unpack(">BBBBBB", data[0:6]),
        "channel":      struct.unpack(">B", data[6])[0],
        "rssi":         struct.unpack(">B", data[7])[0],
        "local_port":   struct.unpack(">H", data[8:10])[0],
        "rtc_value":    struct.unpack(">L", data[10:14])[0],
        "u_bat":        struct.unpack(">H", data[14:16])[0],
        "gpio_pins":    struct.unpack(">H", data[16:18])[0],
        "ascii_time":   struct.unpack(">13s", data[18:31])[0],
        "version":      struct.unpack(">26s", data[32:58])[0],
        "device_id":    struct.unpack(">32s", data[60:92])[0],
        "boot_time":    struct.unpack(">H", data[92:94])[0],
        "sensors":      struct.unpack(">HHHHHHHH", data[94:110])
    }

    info["ap_mac"] = ":".join(["%02x" % x for x in info["ap_mac"]])

    # The string fields are zero-terminated
    #
    info["ascii_time"] = string.split(info["ascii_time"], "\0")[0]
    info["version"] = string.split(info["version"], "\0")[0]
    info["device_id"] = string.split(info["device_id"], "\0")[0]

    # Timestamp to filter out old entries
    #
    info["last_seen"] = datetime.now()
    return info


class WiFlyListener(QObject):
    def __init__(self):
        QObject.__init__(self)

        # List of clients
        #
        self.clients = {}

        # Open socket to receive UDP broadcasts
        #
        self.socket = QUdpSocket(self)
        self.socket.bind(BROADCAST_PORT, QUdpSocket.ShareAddress)
        self.socket.readyRead.connect(self.update)

    def update(self):
        while self.socket.hasPendingDatagrams():
            (data, host, port) = self.socket.readDatagram(4096)
            host = host.toString()

            try:
                self.clients[host] = unpack_info(data)

            except struct.error as exc:
                # Ignore non-WiFly messages
                #
                print "Can't unpack WiFly broadcast message from %s: %s" % (host, str(exc))
                pass

File: Tools/quadcontrol/WiFlyListener.cpp
/**
 * UDP broadcast receiver for Roving Networks/Microchip WiFly Modules
 *
 * Copyright (C)2015 Thomas Kindler <mail_drquad@t-kindler.de>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "WiFlyListener.h"
#include <QtEndian>


#pragma pack(push, 1)
struct BroadcastPacket {
    uint8_t     apMacAddress[6];    // 0-5
    uint8_t     channel;            // 6
    uint8_t     rssi;               // 7
    uint16_t    be_localPort;       // 8
    uint32_t    be_rtcValue;        // 10
    uint16_t    be_uBat;            // 14
    uint16_t    be_gpioPins;        // 16
    char        asciiTime[13+1];    // 18
    char        version[26+2];      // 32
    char        deviceId[32];       // 60
    uint16_t    be_bootTime;        // 92
    uint16_t    be_sensors[8];      // 94
};
#pragma pack(pop)


inline bool qMapLessThanKey(const QHostAddress &key1, const QHostAddress &key2)
{
    return key1.toString() < key2.toString();
}


WiFlyListener::WiFlyListener(QObject *parent)
    : QObject(parent)
    , socket(this)
{
    socket.bind(BROADCAST_PORT, QUdpSocket::ShareAddress);
    connect(&socket, &QUdpSocket::readyRead, this, &WiFlyListener::socket_readyRead);
}


WiFlyListener::~WiFlyListener()
{
}


bool WiFlyListener::handlePacket(const QByteArray &packet, const QHostAddress &addr)
{
    if (packet.size() != sizeof(BroadcastPacket))
        return false;

    auto &p = *(const BroadcastPacket *)packet.constData();

    ClientInfo info;

    info.lastSeen = QDateTime::currentDateTime();

    info.apMacAddress.sprintf(
        "%02x:%02x:%02x:%02x:%02x:%02x",
        p.apMacAddress[0], p.apMacAddress[1],
        p.apMacAddress[2], p.apMacAddress[3],
        p.apMacAddress[4], p.apMacAddress[5]
    );

    info.channel    = p.channel;
    info.rssi       = p.rssi;
    info.localPort  = qFromBigEndian(p.be_localPort);
    info.rtcValue   = qFromBigEndian(p.be_rtcValue);
    info.uBat       = qFromBigEndian(p.be_uBat);
    info.gpioPins   = qFromBigEndian(p.be_gpioPins);

    info.asciiTime.sprintf("%.*s", (int)sizeof(p.asciiTime), p.asciiTime);
    info.version.sprintf("%.*s", (int)sizeof(p.version), p.version);
    info.deviceId.sprintf("%.*s", (int)sizeof(p.deviceId), p.deviceId);
    info.bootTime_ms = qFromBigEndian(p.be_bootTime);

    for (int i=0; i<8; i++)
        info.sensors[i] = qFromBigEndian(p.be_sensors[i]);

    clients[addr] = info;

    return true;
}


void WiFlyListener::socket_readyRead()
{
    while (socket.hasPendingDatagrams()) {
        QByteArray      packet;
        QHostAddress    addr;

        packet.resize(socket.pendingDatagramSize());
        socket.readDatagram(packet.data(), packet.size(), &addr);

        // qDebug(
        //     "WiFly message from %s: %s",
        //     qPrintable(addr.toString()),
        //     qPrintable(packet.toHex())
        // );

        handlePacket(packet, addr);
    }
}
